### PR TITLE
feat(refactor): ExtractCall engine (hoist resolve block into spec.calls)

### LINKED ISSUE #778: feat(refactor): ExtractCall engine (hoist resolve block into spec.calls)
> **Assignment:** Dev B (independent-features stream) | **Phase 1**
> **Blocked by:** nothing (pure `pkg/refactor`) | **Blocks:** #779
> **Sync point:** none -- zero-dependency parallel starter; begin on day 1
> alongside #770. No LSP/server contention.

**PR unit:** R7a | **Complexity:** L | **Depends on:** nothing new (pure `pkg/refactor`)
**Parallel starter:** can begin during foundation; blocks #779 (R7b).

## Scope

Add a domain-level refactoring engine that hoists a selected `resolve`/step block
into a reusable `spec.calls` definition and rewrites the selected usage to a
`call:` reference. This is the only substantial NEW domain logic in the roadmap;
it lives in `pkg/refactor`, NOT `pkg/lsp`.

### v1 scope (locked decision -- matches IntelliJ "Extract Method")

- **Selection-driven:** extract the single block the user selected.
- After extracting, **offer** to replace *identical* occurrences elsewhere with
  the new `call:` (opt-in), but do NOT auto-scan for near-duplicates or mass
  rewrite across the solution. (A future "duplicate resolve block" lint rule can
  feed this same engine.)
- Preserve comments/formatting (reuse the existing `pkg/refactor` node-edit
  approach used by rename; `refactor.go` `RenameResult`/`Apply`).
- Infer call `args` from values in the block that vary vs. are literal (keep v1
  conservative: literal block -> no args, or a documented simple inference).

### Interface (contract for #779)

~~~go
func ExtractCall(sol *solution.Solution, blockPath string, callName string) (*RenameResult, error)
// RenameResult.Apply(raw) -> edited bytes, reusing existing apply machinery.
~~~

## Files

- New: `pkg/refactor/extract.go` (+ `extract_test.go`, `extract_benchmark_test.go`)

## Tests / docs

- Table tests: extract a resolver step -> new call + `call:` reference; identical
  occurrence replacement (opt-in); comment/format preservation; error on invalid
  path / name collision.
- Benchmark (new engine path).
- Doc: mention in `docs/design/cli.md` refactor section; a `scafctl refactor
  extract-call` CLI subcommand MAY be added here or deferred (note which).

## Acceptance criteria

- [ ] Extracting a block produces a valid `spec.calls` entry + rewritten usage.
- [ ] Identical-occurrence replacement is opt-in, not automatic.
- [ ] Comments/formatting preserved; output lints clean.
- [ ] Engine is pure domain code (no LSP/protocol imports).

### PR DESCRIPTION
**Pull request by kcloutie**
## Summary

Adds a domain-level **Extract Call** refactoring engine in `pkg/refactor` that hoists a selected resolve/transform/validate step (a `with[i]` provider+inputs block) out of its resolver into a reusable `spec.calls` definition, and rewrites the selected step to a `call: <name>` reference. This is the substantial new domain logic that backs the future LSP code action (#779).

## Contract (for #779)

```go
func ExtractCall(sol *solution.Solution, blockPath, callName string) (*RenameResult, error)
func ExtractCallReplacingIdentical(sol *solution.Solution, blockPath, callName string) (*RenameResult, error)
```

Both return the existing `RenameResult`, so callers reuse `RenameResult.Apply(raw)`.

## Changes

- **`pkg/refactor/extract.go`** -- the engine. Source-preserving byte edits (no YAML round-trip): the step's provider/inputs and inline comments are spliced verbatim and re-indented to the file's style; comments/formatting elsewhere are untouched. All-or-nothing like `RenameSymbol` (any problem returns an error and no edits). **Pure domain code -- no LSP/protocol imports** (verified via `go list -deps`).
- **v1 scope (conservative):** only direct provider steps are extractable (a step already using `call:` is rejected); no argument inference (inputs reproduced literally); the base `ExtractCall` never scans/mass-rewrites. Only `provider`/`inputs` steps are extractable: a step is hoisted verbatim into a `spec.calls` entry, whose type (`spec.Call`) cannot model step-level fields, so a step carrying `when`, `continueOnError`, `onError`, `forEach`, or a validation `message` is **rejected** rather than silently dropping that field (which would splice it into the call body, where it is ignored, and strip it from the call site -- changing conditional/iteration/error-handling behavior with no error). The opt-in `ExtractCallReplacingIdentical` additionally rewrites every *structurally identical* step (deep value equality, key-order/format independent) into the same `call:` reference -- never near-matches.
- **Robustness:** rejects an ambiguous empty-but-present `spec.calls` (inline `{}` or a bare/comment-only key) rather than emitting a duplicate `calls:` key, and rejects a flow-style insertion target (an existing `calls: {a: {...}}` or a flow-style `spec: {...}`) rather than splicing a block entry into a flow mapping and corrupting the document. Block end-span is computed by indentation scanning (yaml.v3 gives only the node start); the new `calls:` insertion point is deterministic (append after the last existing entry, or a new `calls:` as the first child of `spec`), with indentation derived from real node positions.
- **`docs/design/cli.md`** -- documents the engine; the `scafctl refactor extract-call` CLI subcommand is **deferred** to a later PR.

## Tests / verification

- `pkg/refactor/extract_test.go`: happy path, transform step, append-to-existing-calls, middle-of-list sibling survival, identical-occurrence replacement, near-match exclusion, all error cases, an empty-`calls` regression, an unsupported-step-field regression (`when`/`continueOnError`/`forEach`/`message` are rejected), and a flow-style regression (flow-style `spec.calls`/`spec` are rejected). Every rewrite is re-parsed and `ValidateSpec()`-checked.
- `pkg/refactor/extract_benchmark_test.go`: `BenchmarkExtractCall` (~32us/op).
- `go build ./...`, `go vet ./...`, `go test ./pkg/refactor/...`, `go test ./tests/integration/...` -- all pass.
- `task lint:changed` -- 0 new issues.
- Reviewed by a dedicated code-review pass; a file-corruption edge case (empty `spec.calls`) it found was fixed and regression-tested before opening. Follow-up review passes found (a) steps with unsupported step-level fields were hoisted verbatim (silently losing `when`/`continueOnError`/`forEach`/`message`), and (b) a flow-style `spec.calls`/`spec` produced a corrupt (unparseable) rewrite; both are now rejected with a clear error and regression-tested.

## Acceptance criteria

- [x] Extracting a block produces a valid `spec.calls` entry + rewritten usage
- [x] Identical-occurrence replacement is opt-in, not automatic
- [x] Comments/formatting preserved; output re-parses and validates
- [x] Engine is pure domain code (no LSP/protocol imports)

Closes #778

### CODE REVIEW

File: pkg/refactor/extract.go [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+	li := sourcepos.NewLineIndex(raw, "")
+	sm := sol.SourceMap()
+
+	// A spec.calls key present in source but with no decodable entries (an inline
+	// "calls: {}", or a bare/comment-only "calls:") is ambiguous: HasCalls() is
+	// false, so the insertion logic below would create a SECOND "calls:" key and
+	// corrupt the file. Refuse rather than emit a broken edit set (all-or-nothing).
+	if _, present := sm.Get("spec.calls"); present && !sol.Spec.HasCalls() {
+		return nil, fmt.Errorf("extract call: spec.calls is present but has no entries (empty or inline); populate or remove it before extracting")
+	}
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/refactor/extract.go:121]: Thanks -- verified this is already safe, so leaving the code as-is (non-blocking).

`SourceMap.Get` guards a nil receiver (`pkg/sourcepos/sourcepos.go:66-69`: `if sm == nil { return Position{}, false }`), and Go permits method calls on a nil pointer, so `sm.Get("spec.calls")` cannot panic even when `sol.SourceMap()` returns nil -- it returns `(Position{}, false)`. In that case the downstream `callIndents`/`callsInsertion` lookups miss `spec` and return a clear `"cannot locate spec block"` error with no edits, which is exactly the "fail with a clear error instead of dereferencing nil" behavior requested. Resolving as a non-blocking nit.

File: pkg/refactor/extract.go [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+		}
+	}
+	if hasCall {
+		return fmt.Errorf("step already uses a call reference; only direct provider steps are extractable")
+	}
+	if !hasProvider {
+		return fmt.Errorf("step is not a provider step (no provider key); only direct provider steps are extractable")
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/refactor/extract.go:238]: Good catch -- this was a real silent-behavior-loss bug, fixed in e4d66b1e.

Confirmed empirically: a step with `provider` + `inputs` + `when` + `continueOnError` extracted without error, splicing `when`/`continueOnError` into the `spec.calls` body (where `spec.Call` silently ignores them) and stripping them from the call site -- dropping the resolver's conditional/error-handling behavior.

`requireProviderStep` now rejects any step declaring a field other than `provider`/`inputs` (`call` keeps its own clearer message), covering `when`, `continueOnError`, `onError`, `forEach`, and validation `message`, with a clear error rather than a silent rewrite. Matches v1's conservative scope. Regression tests added in `TestExtractCall_RejectsUnsupportedStepFields`.
